### ISSUE TITLE
Should anonymous functions get a `name` data property?

### ISSUE THREAD

**Issue by ajklein** [tc39/ecma262#1049]
ES2015 and later give a `name` property to named functions (whether named via declaration, expression, or name inference). Anonymous functions aren't given one, presumably with the expectation that `name` lookups will read through to `Function.prototype.name` (which is an empty string).

Yet engines have legacy behavior here:

```
> eshost -e '(() => {}).hasOwnProperty("name")' // same result for function(){}
#### v8
true

#### jsc
true

#### spidermonkey
true

#### chakra
false
```

I'd propose that there'd be little downside to specifying this behavior.

**Comment by ljharb** [comment]
What are the same eshost results for a non-arrow function, a generator function, and an `async function`?

**Comment by ajklein** [comment]
@ljharb all the same as the arrow function above: v8/jsc/spidermonkey all add a `name` property (with the value `""`).

**Comment by ljharb** [comment]
That strongly supports making it an own property, indeed.

**Comment by getify** [comment]
Just out of curiosity, why would engines have chosen, violating the spec, to add a name property (with an actual value like `""` and not even `undefined`) to every single anonymous function (probably thousands in most apps) and yet resist (for performance reasons) adding a `groups` property to every regexp result object? I don't understand why in one case this would be preferred and in another case, not?

**Comment by allenwb** [comment]
Is there an reason why implementations deviated from the spec? If so, what is it?

There was a reason for why it was spec'ed the way it is. If there is a strong implementation reason for the deviation then it would make sense to change the spec.  But changing it just to match unjustified implementations deviations is bad precedent.

**Comment by syg** [comment]
@ajklein What about unnamed class expressions across the engines? I see code in SM explicitly not defining `.name` as an own property on unnamed class exprs.

**Comment by ajklein** [comment]
@allenwb I believe the engines which return `true` above all implemented a `name`  property before it was added to ECMAScript in version 6. For V8, that's definitely the case; see the [tracking bug](https://bugs.chromium.org/p/v8/issues/detail?id=4709) for more background. From that point of view, ES2015 is the "unjustified" one in deviating from existing behavior.

@syg That's very interesting: only JSC provides a `name` property for an anonymous class.

**Comment by ljharb** [comment]
@allenwb could you elaborate on the reasons why this is spec'ed this way?

**Comment by allenwb** [comment]
@ajklein At the time `name` was added to ES6 it  non-standard and wasn't universally  implemented by major browsers (at least IE did not implement it).  I'm sure there was variation among those that did implement it, just as has been reported in this thread.  Given that situations, TC39 evaluated the utility of the feature, decided it was useful, looked at specification alternatives, made some decisions, and wrote the spec.  All the major implementations were represented in that  process and ultimately voted to approved the final ES6 spec.  If there were objections to ES6's choosen semantics, there were plenty of opportunity to raise objections.   Given all that, this sounds to me like a set of implementation bugs where implementations have not been updated to match the approved standard.

Note, that the actual design point here is not a very big deal.  It could have gone either way.  But the whole point of having a standard is to pin down such design points so implementations don't just go off and follow their preferences.

@ljharb  Why? Because an own `name` property whose value is "" is redundant since such a property is inherited from Function.prototype. Plus there are many anonymous function instances .  All those redundant properties add up.

**Comment by ljharb** [comment]
Thanks for clarifying, that's helpful.

**Comment by annevk** [comment]
I think unless implementations were confronted with the fact that they would have to remove functionality (and bugs were filed) it's hard to hold them accountable. It's rather easy for those kind of things to go unnoticed (as seems to have happened here). When there's existing legacy behavior you have to go the extra mile to get interoperability.

**Comment by allenwb** [comment]
@annevk Yes, these sort of things can be missed. During the ES6 period  it was really the responsibility of the TC39 representatives of various implementations to key their implementors up to date on changes and it didn't always happen.

**Comment by TimothyGu** [comment]
Marginally related: In addition to merely having a `name` own property, at least SpiderMonkey and V8's Proxy revocation functions have a `name` property that is `"revoke"` rather than the empty string. This seems like oversights, and thus filed at

- https://crbug.com/v8/7245
- https://bugzilla.mozilla.org/show_bug.cgi?id=1427058

**Comment by ajklein** [comment]
@allenwb I agree it's great to have a standard to pin down design points (and we're on the same page that this particular design choice isn't particularly important one way or the other, all things being equal). But I object to disallowing changes to those designs solely by pointing to the ratification of ES6 (especially given the size of that document and its long gestation period).

**Comment by ljharb** [comment]
With latest engines, when I run `eshost test.js`:
<details>
<summary>`test.js` contents</summary>

```js
function getLoc(src) {
	var key = 'name';
	var obj;
	try {
		obj = Function('return ' + src)();
	} catch (e) {
		return 'unsupported';
	}
	if (Object.prototype.hasOwnProperty.call(obj, key)) {
		return 'own';
	}

	var protos = 0;
	var O = obj.__proto__;
	while (O) {
		protos++;
		if (Object.prototype.hasOwnProperty.call(O, key)) {
			return 'proto: ' + protos;
		}
		O = obj.__proto__;
		if (protos > 20) {
			return 'cycle somewhere?';
		}
	}
	return 'none';
}

print(JSON.stringify({
	'Function.prototype': getLoc('Function.prototype'),
	'anon func': getLoc('function () {}'),
	'arrow func': getLoc('() => {}'),
	'anon class': getLoc('class {}'),
	'anon gen': getLoc('function* () {}'),
	'anon async': getLoc('async function () {}'),
	'anon async gen': getLoc('async function* () {}'),
}, null, '\t'));
```
</details>

Results:
```js
#### Chakra
{
	"Function.prototype": "own",
	"anon func": "proto: 1",
	"arrow func": "proto: 1",
	"anon class": "proto: 1",
	"anon gen": "cycle somewhere?",
	"anon async": "cycle somewhere?",
	"anon async gen": "unsupported"
}

#### JavaScriptCore
{
	"Function.prototype": "own",
	"anon func": "own",
	"arrow func": "own",
	"anon class": "own",
	"anon gen": "own",
	"anon async": "own",
	"anon async gen": "own"
}

#### SpiderMonkey
{
	"Function.prototype": "own",
	"anon func": "own",
	"arrow func": "own",
	"anon class": "proto: 1",
	"anon gen": "own",
	"anon async": "own",
	"anon async gen": "own"
}

#### V8
{
	"Function.prototype": "own",
	"anon func": "own",
	"arrow func": "own",
	"anon class": "proto: 1",
	"anon gen": "own",
	"anon async": "own",
	"anon async gen": "own"
}
```

(really not sure about the apparent cycles in chakra for anonymous generators and anonymous async functions)

Only anonymous classes seem to follow the spec in 3/4 browsers; with everything else, it's an own property in every browser that supports the construct.

Of the rationales explained [above](https://github.com/tc39/ecma262/issues/1049#issuecomment-352961622), it seems that implementations are not concerned about the number of properties "adding up", but the redundancy is a solid point.

Web reality currently says it's an own property - the choices seem to be either:
1. change the spec to match web reality, with an exception for anonymous classes
1. change the spec to match web reality, but pursuing implementations changing anonymous classes to have an own property
1. pursue implementations changing things so that the spec remains correct (although everyone would still have to change anonymous classes to *have* an own property)

What are the thoughts of implementors here? Is there a preference? Note that there is no option that  both involves no work from implementors and also leaves the spec in a consistent state (ie, where all anonymous functions either have an own property, or lack it).

**Comment by ajklein** [comment]
With my implementor hat on, either (1) or (2) is much-preferable to (3), mostly to avoid churn. I don't have a strong preference between (1) and (2) (though I see the consistency argument).

**Comment by anba** [comment]
> Of the rationales explained [above](https://github.com/tc39/ecma262/issues/1049#issuecomment-352961622), it seems that implementations are not concerned about the number of properties "adding up", but the redundancy is a solid point.

I can't respond for other implementations, but at least as far as SpiderMonkey is concerned, the "name" (and also the "length" and "prototype") property is added lazily for functions objects in SpiderMonkey. This avoids to create these properties eagerly, resulting in performance and memory wins. 

There are also a number of downsides of this approach: It messes up property enumeration order, requires additional book-keeping, and leads to reshaping objects affecting ICs, see <https://bugzilla.mozilla.org/show_bug.cgi?id=1237935>. From the spec-conformance POV, the enumeration order issue is only relevant. 

```
function A() {}
A.foo = 0;

// Should print "foo" as the last property, but is the first one in SpiderMonkey. 
print(Reflect.ownKeys(A));
```

In addition to lazily creating the properties, SpiderMonkey is also reusing the book-keeping mechanism needed for the lazy creation hook for further optimizations. For example for the "name" property, there's a flag to signal when the "name" was actually defined as a property on the function object. So when the flag is not yet set, SpiderMonkey can retrieve the function's name directly from a field of the C++ object instead of performing a property lookup. This optimization is for example used when creating bound functions, because it avoids performing a property access for the "name" of the bound function target in the implementation of [19.2.3.2 Function.prototype.bind](https://tc39.github.io/ecma262/#sec-function.prototype.bind), steps 9-11, and by that makes it possible to more easily create [assembly code](https://searchfox.org/mozilla-central/rev/330daedbeac2bba296d663668e0e0cf248bc6823/js/src/jit/CodeGenerator.cpp#13379-13385) for these steps.

Basically all that means it doesn't seem likely at this point in time to see any kind of effort (in SpiderMonkey) to ensure anonymous functions are created without a "name" property.

Or to be more correct: To ensure a "name" property is not *lazily* added to anonymous functions. :smile:

**Comment by littledan** [comment]
It seems like we've gotten to a good point in this discussion, where the next step would be to write a PR for either option 1. or 2. from https://github.com/tc39/ecma262/issues/1049#issuecomment-452923011 . (Personally, I don't have a strong preference among those.)

**Comment by ljharb** [comment]
I have a mild preference for option 2, because it seems more consistent.

**Comment by jorendorff** [comment]
We'd support option 2. Sounds good.
